### PR TITLE
[Feature] Add Resnet2dAudio

### PR DESCRIPTION
**Pull request by SuX97**
This PR adds
+ code and config AudioOnly Pathway for AVSlowFast and its checkpoint
+ ~~code and config for AVSlowfast backbone~~

### CODE REVIEW

File: mmaction/models/common/conv_audio.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    Args:
+        in_channels (int): Same as nn.Conv3d.
+        out_channels (int): Same as nn.Conv3d.
+        kernel_size (int | tuple[int]): Same as nn.Conv3d.
+        stride (int | tuple[int]): Same as nn.Conv3d.
+        padding (int | tuple[int]): Same as nn.Conv3d.
+        dilation (int | tuple[int]): Same as nn.Conv3d.
+        groups (int): Same as nn.Conv3d.
+        bias (bool | str): If specified as ``auto``, it will be decided by the
```
Comment (dreamerlin) [inline mmaction/models/common/conv_audio.py:18]:
```text
`auto` rather than ``auto``, since it is a str not a variable
```

File: mmaction/models/common/conv_audio.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    """Conv2d module for AudioResNet backbone.
+
+    Args:
+        in_channels (int): Same as nn.Conv3d.
+        out_channels (int): Same as nn.Conv3d.
+        kernel_size (int | tuple[int]): Same as nn.Conv3d.
+        stride (int | tuple[int]): Same as nn.Conv3d.
+        padding (int | tuple[int]): Same as nn.Conv3d.
+        dilation (int | tuple[int]): Same as nn.Conv3d.
+        groups (int): Same as nn.Conv3d.
```
Comment (dreamerlin) [inline mmaction/models/common/conv_audio.py:17]: 1. missing default value.
2. https://github.com/open-mmlab/mmcv/pull/651

File: mmaction/models/backbones/resnet_audio.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+        else:
+            out = _inner_forward(x)
+
+        out = self.relu(out)
+
+        return out
+
+
+@BACKBONES.register_module
+class ResNet2dAudio(nn.Module):
```
Comment (dreamerlin) [inline mmaction/models/backbones/resnet_audio.py:116]: How about inheriting from `ResNet`

Comment (SuX97) [reply mmaction/models/backbones/resnet_audio.py:116]: Yeah, I have considered it. Actually, it is more like a ResNet3D. It needs to factorize all conv2ds in the first two res-stages, which is similar to ResNet3D(which inflates some stages). Almost all functions should be overrides including BasicBlock/Bottleneck, make_res_stage, init_weight. I think it's better to write it independently like ResNet3D for clarity.

File: configs/recognition_audio/avslowfast/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+}
+```
+
+## Model Zoo
+
+### Kinetics-400
+
+|config | n_fft | gpus | backbone |pretrain| top1 acc/delta| top5 acc/delta | inference_time(video/s) | gpu_mem(M)| ckpt | log| json|
+|:--|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|
+|[avslowfast_r50_32x2x1_239e_kinetics400_audio_feature](/configs/recognition_audio/avslowfast/avslowfast_r50_32x2x1_239e_kinetics400_audio_feature.py)|1024|8| ResNet50 | None |||x||[ckpt]()|[log]()|[json]()|
```
Comment (dreamerlin) [inline configs/recognition_audio/avslowfast/README.md:19]: Is the ckpt ready?

File: configs/recognition_audio/resnet/tsn_resnet_r18_64x1x1_100e_kinetics400_audio_feature.py [hunk 5]
```diff
@@ -90,7 +90,7 @@
 # runtime settings
 dist_params = dict(backend='nccl')
 log_level = 'INFO'
-work_dir = './work_dirs/tsn_r18_64x1x1_100e_kinetics400_audio_feature/'
+work_dir = './work_dirs/tsn_resnet_r18_64x1x1_100e_kinetics400_audio_feature/'
```
Comment (dreamerlin) [inline configs/recognition_audio/resnet/tsn_resnet_r18_64x1x1_100e_kinetics400_audio_feature.py:93]: why indicating `resnet_r18` not just `r18`

File: mmaction/datasets/audio_feature_rawframe_dataset.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+from .registry import DATASETS
+
+
+@DATASETS.register_module
+class AudioFeatureRawframeDataset(RawframeDataset):
+
+    def __init__(self, ann_file, pipeline, audio_prefix, **kwargs):
+        self.audio_prefix = audio_prefix
+        super(AudioFeatureRawframeDataset,
+              self).__init__(ann_file, pipeline, **kwargs)
```
Comment (dreamerlin) [inline mmaction/datasets/audio_feature_rawframe_dataset.py:13]: ~AudioF...~

File: mmaction/models/backbones/resnet3d_avslowfast.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+                kernel_size=(fusion_kernel, 1),
+                stride=(self.speed_ratio, 1),
+                padding=((fusion_kernel - 1) // 2, 0),
+                bias=False,
+                conv_cfg=self.conv_cfg,
+                norm_cfg=self.norm_cfg,
+                act_cfg=None)
+
+        self.lateral_connections = []
+        for i in range(len(self.stage_blocks)):
```
Comment (dreamerlin) [inline mmaction/models/backbones/resnet3d_avslowfast.py:60]: pylint recommends `for i, _ in enumerate(self.stage_blocks)`

File: mmaction/models/backbones/resnet3d_avslowfast.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        nn.Module: Created pathway.
+    """
+    if not (isinstance(cfg, dict) and 'type' in cfg):
+        raise TypeError('cfg must be a dict containing the key "type"')
+    cfg_ = cfg.copy()
+
+    pathway_type = cfg_.pop('type')
+    if pathway_type not in pathway_cfg:
+        raise KeyError(f'Unrecognized pathway type {pathway_type}')
+    else:
```
Comment (dreamerlin) [inline mmaction/models/backbones/resnet3d_avslowfast.py:138]: `no-else-raise` (the `else` is redundant after raise)

File: mmaction/models/backbones/resnet_audio.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+        assert len(factorize) == blocks
+        downsample = None
+        if stride != 1 or inplanes != planes * block.expansion:
+            downsample = ConvModule(
+                inplanes,
+                planes * block.expansion,
+                kernel_size=1,
+                stride=stride,
+                bias=False,
+                norm_cfg=norm_cfg,
```
Comment (dreamerlin) [inline mmaction/models/backbones/resnet_audio.py:254]: last comma is redundant

File: mmaction/models/heads/avslowfast_head.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+        Args:
+            x (torch.Tensor): The input data.
+
+        Returns:
+            torch.Tensor: The classification scores for input samples.
+        """
+        # ([N, channel_fast, T, H, W], [(N, channel_slow, T, H, W)])
+        x_slow, x_fast, x_audio = x
+        if len(x_audio.size()) == 4:
```
Comment (dreamerlin) [inline mmaction/models/heads/avslowfast_head.py:34]: x_audio.dim()

File: mmaction/models/recognizers/audio_visual_recognizer.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        x = self.extract_feat(imgs, audios)
+        cls_score = self.cls_head(x)
+        cls_score = self.average_clip(cls_score)
+
+        return cls_score.cpu().numpy()
+
+    def forward(self, imgs, audios, label, return_loss=True):
+        if return_loss:
+            return self.forward_train(imgs, audios, label)
+        else:
```
Comment (dreamerlin) [inline mmaction/models/recognizers/audio_visual_recognizer.py:57]: no-else-return

File: mmaction/models/backbones/resnet3d_avslowfast.py [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+                     base_channels=8,
+                     conv1_kernel=(5, 7, 7),
+                     conv1_stride_t=1,
+                     pool1_stride_t=1),
+                 audio_pathway=dict(
+                     type='resnet_audio',
+                     depth=50,
+                     pretrained=None,
+                     strides=(2, 2, 2, 2),
+                     lateral=True)):
```
Comment (dreamerlin) [inline mmaction/models/backbones/resnet3d_avslowfast.py:221]: use `copy.deepcopy()` to protect the mutable value as default (the dict)

File: configs/recognition_audio/avslowfast/README.md (hunk 4 above)
Comment (SuX97) [reply configs/recognition_audio/avslowfast/README.md:19]: Still training now, takes more than training a slowfast.

File: configs/recognition_audio/resnet/tsn_resnet_r18_64x1x1_100e_kinetics400_audio_feature.py (hunk 5 above)
Comment (SuX97) [reply configs/recognition_audio/resnet/tsn_resnet_r18_64x1x1_100e_kinetics400_audio_feature.py]: Right. I have re-organized this one. `resnet` and `resnet_audio` denote two different backbones. Now it is called `resnet` and `audioonly` (a similar way as slowonly).

**Comment by SuX97** [comment]
@innerlee Kindly ping, this PR is ready to go.

**Review by innerlee — APPROVED** [review]
